test/e2e/e2e_suite_test.go

```diff
@@ -79,11 +79,19 @@ var _ = BeforeSuite(func() {
 	}
 
 	nroObj := objects.TestNRO(matchLabels)
+	kcObj := objects.TestKC(matchLabels)
+	pauseMCPs(nroObj.Spec.NodeGroups, true)
+
+	By(fmt.Sprintf("creating the KC object: %s", kcObj.Name))
+	err = e2eclient.Client.Create(context.TODO(), kcObj)
+	Expect(err).NotTo(HaveOccurred())
 
 	By(fmt.Sprintf("creating the NRO object: %s", nroObj.Name))
 	err = e2eclient.Client.Create(context.TODO(), nroObj)
 	Expect(err).NotTo(HaveOccurred())
 
+	pauseMCPs(nroObj.Spec.NodeGroups, false)
```

Comment:
next time consider creating two functions:
`func pauseMCPs(nodeGroups []nropv1alpha1.NodeGroup)`
and
`func unpauseMCPs(nodeGroups []nropv1alpha1.NodeGroup)`
The reason is that the bool argument feels clunky. This approach may be more expressive and we can share quite a lot of common code anyway

Reply:
Initially, I went with this approach but then I saw the functions are 100% the same but the ```pause``` field.
So I thought it's a huge waste of repetitive code.

Reply:
What I'd probably do is something like
```
func pauseMCPs(nodeGroups []nropv1alpha1.NodeGroup) func() {
	mcps, err := controllers.GetNodeGroupsMCPs(context.TODO(), e2eclient.Client, nodeGroups)
	Expect(err).NotTo(HaveOccurred())
	Expect(len(mcps).To(BeGreatherThan(0).ICantRememberThe(Syntax).Please(Check))
	for i := range mcps {
		mcps[i].Spec.Paused = true
		err = e2eclient.Client.Update(context.TODO(), mcps[i])
		Expect(err).NotTo(HaveOccurred())
	}
	unpause := func() {
		for i := range mcps {
			mcps[i].Spec.Paused = false
			err = e2eclient.Client.Update(context.TODO(), mcps[i])
			Expect(err).NotTo(HaveOccurred())
		}
	}
	return unpause
}
```
Use like

unpause := pauseMCPs(nroObj.Spec.NodeGroups)

// do stuff

unpause()
```

the advantage is we make sure to unpause the same set of MCPs we pause. It's unlikely some MCPs get created in between, but from theoretical perspective this is pretty clean and not too clunky. Note this is a pretty common golang/kube pattern (returning an anonymous function which does cleanup), not inventing anything here.

Another option is to save the MCPs we pause. Not strong feelings against that, it's ok as well.

Reply:
if we go the fancy way with the anonymous func, let's make sure we capture the MCPs properly in the closure, this is a common golang bug.

Reply:
This is such a neat trick! I'll totally adopt it.

> let's make sure we capture the MCPs properly in the closure

What do you mean by that, from your code example it looks like we're working on the same group of MCPs so it looks ok to me

Reply:
We need to make sure we don't introduce hard-to see bugs caused to shadowing. `for` loops and closures are the usual suspects for this kind of bugs. Code looks correct, yet it misbehaves. It's one of the most common golang gotchas. See for example: https://blog.charmes.net/post/scope-shadowing-go/

Reply:
that said, it is possible my code is actually correct, but I didn't spent much thought on it, so it must be taken with a big grain of salt.